File: src/main/scala/saiml/nn/perceptron/Perceptron.scala
package saiml.nn.perceptron

import saiml.math.VectorOp._

/**
  * A very simple single-layer perceptron model.
  *
  * Ref.: https://en.wikipedia.org/wiki/Perceptron
  */
class Perceptron(val weights: Vector[Double], val bias: Double = 0.0) {

  /** Activation function */
  def f(input: Vector[Double]): Boolean = weights * input + bias > 0

  def learn(input: Vector[Double], target: Boolean): Perceptron = {
    require(input.size == weights.size, "input size mismatch")

    // Internally uses vector size one larger, always sets input(0) to 1.
    // Then, weights(0) is effectively the learned bias.
    val p = new Perceptron(bias +: weights)
    val learnInput = 1.0 +: input

    val outNum: Double = if (p.f(learnInput)) 1 else 0
    val targetNum: Double = if (target) 1 else 0

    val newWeights = p.weights zip learnInput map Function.tupled { (w, x) =>
      w + (targetNum - outNum) * x
    }

    new Perceptron(weights = newWeights.tail, bias = newWeights.head)
  }

  /** Convenience shortcut for feeding several examples */
  def learn(examples: Traversable[(Vector[Double], Boolean)]): Perceptron = {
    examples.foldLeft(this) { (acc, x) =>
      acc.learn(x._1, x._2)
    }
  }
}

File: swiftlearner_cpp/src/nn/Perceptron.cpp
#include "nn/Perceptron.hpp"
#include <numeric> // std::inner_product

/**
  * A very simple single-layer perceptron model.
  *
  * Ref.: https://en.wikipedia.org/wiki/Perceptron
  */
Perceptron::Perceptron(const vec& weights, double bias) : weights_(weights), bias_(bias) {}

bool Perceptron::activate(const vec& input) const {
  if (input.size() != weights_.size())
    throw std::invalid_argument("input size mismatch");

  return std::inner_product(weights_.begin(), weights_.end(), input.begin(), bias_) > 0;
}

Perceptron Perceptron::learn(const vec& input, bool target) {
  if (input.size() != weights_.size())
    throw std::invalid_argument("input size mismatch");

  double output = activate(input) ? 1.0 : 0.0;
  double error = (target ? 1.0 : 0.0) - output;

  bias_ += error * 1.0; // Input is implicitly 1 for bias update

  for (size_t i = 0; i < weights_.size(); ++i)
    weights_[i] += error * input[i];

  return *this;
}

/** Convenience shortcut for feeding several examples */
Perceptron Perceptron::train(const std::vector<std::pair<vec, bool>>& examples, int epochs) {
  for (int i = 0; i < epochs; ++i)
    for (const auto& [input, target] : examples)
      *this = learn(input, target);
  return *this;
}